### PR TITLE
Part 3 of 3: DetailsList CSS-in-JS conversion

### PR DESCRIPTION
**Pull request by kenotron**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

This is a relatively small final checkin to convert the last bits of detailslist to use mergestyles.

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by kenotron** [comment]
@JasonGore The last one! This one is significantly smaller than the rest of the DL conversion PRs

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 1]
```diff
@@ -7,5 +7,7 @@ export { IDetailsHeaderProps };
 
 export const DetailsHeader = styled<IDetailsHeaderProps, IDetailsHeaderStyleProps, IDetailsHeaderStyles>(
   DetailsHeaderBase,
-  getStyles
+  getStyles,
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: Rename to DetailsHeader.ts (same thing for the other files like this)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx [hunk 2]
```diff
@@ -1,968 +1,15 @@
-import * as React from 'react';
-import * as stylesImport from './DetailsList.scss';
+import { styled } from '../../Utilities';
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx]: If you did a rename of DetalisList.tsx -> DetailsList.base.tsx in a commit,, then create a brand new DetailsList.ts (not tsx) with the styled export, the line history gets preserved and you don't show up on gitblame for the entire file :p

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 3]
```diff
@@ -41,7 +42,11 @@ export interface IDetailsList extends IList {
   getStartItemIndexInView: () => number;
 }
 
-export interface IDetailsListProps extends React.Props<DetailsList>, IWithViewportProps {
+export interface IDetailsListProps extends React.Props<DetailsListBase>, IWithViewportProps {
+  theme?: ITheme;
+
+  styles?: IStyleFunctionOrObject<IDetailsListStyleProps, IDetailsListStyles>;
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Can we please document both styles and themes? (I'm sure you can find examples of components which do not, but most do - and we really ought to be documenting)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 4]
```diff
@@ -501,3 +506,15 @@ export enum CheckboxVisibility {
    */
   hidden = 2
 }
+
+export type IDetailsListStyleProps = Required<Pick<IDetailsListProps, 'theme'>> &
+  Pick<IDetailsListProps, 'className'> & {
+    isHorizontalConstrained?: boolean;
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: We really should be documenting these as well, if you think about the purpose these props serve and from the POV of someone writing a style function for DetailsList

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 5]
```diff
@@ -501,3 +506,15 @@ export enum CheckboxVisibility {
...
+    compact?: boolean;
+    isFixed?: boolean;
+  };
+
+export interface IDetailsListStyles {
+  root: IStyle;
+  focusZone: IStyle;
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: And arguably these too (again there are various examples of components which do complete documentation). But minimally the styleprops.

(Look at OverflowSet.types.ts for example...)

File: packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.test.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+describe('ShimmeredDetailsList', () => {
+  it('renders List correctly', () => {
+    ShimmeredDetailsList.prototype.componentDidMount = jest.fn();
+
+    const component = renderer.create(
+      <ShimmeredDetailsList
+        items={mockItems(5)}
+        // tslint:disable-next-line:jsx-no-lambda
+        onRenderRow={() => null}
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.test.tsx]: Why not just use jest.fn() in places like this?

File: common/changes/office-ui-fabric-react/detailslist_2018-07-11-05-34.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Remaining CSS-in-JS conversion of DetailsList",
```
[the remark this answers is not in the document]
Comment (kenotron) [reply common/changes/office-ui-fabric-react/detailslist_2018-07-11-05-34.json]: done

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx (hunk 1 above)
Comment (kenotron) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: Done renaming

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts (hunk 4 above)
Comment (kenotron) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: added more docs to these

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts (hunk 5 above)
Comment (kenotron) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: I'm documenting the styleprops for now - DL subcomponents are not expected to be used directly by others. They're actually kind of a distraction. There are only two styles here and they seem to be pretty self-explanatory.

File: packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.test.tsx (hunk 6 above)
Comment (kenotron) [reply packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.test.tsx]: I just cloned this from DL's test... didn't seem to be a big deal to do a jest.fn() or have props be passed in, in this case

Comment (cliffkoh) [reply packages/office-ui-fabric-react/src/components/DetailsList/ShimmeredDetailsList.test.tsx]: Agree it's not a big deal - but you can avoid the tslint disable this way :) Again, no big deal.

**Pull request merged**